common/locks/id_mutex.go

```diff
@@ -32,17 +32,20 @@ type (
 	// HashFunc represents a hash function for string
 	HashFunc func(interface{}) uint32
 
+	// UnlockFunc unlocks a lock
+	UnlockFunc func()
+
 	// IDMutex is an interface which can lock on specific comparable identifier
 	IDMutex interface {
-		LockID(identifier interface{})
-		UnlockID(identifier interface{})
+		LockID(identifier interface{}) UnlockFunc
 	}
 
 	// idMutexShardImpl is the implementation of IDMutex shard
 	idMutexImpl struct {
 		numShard uint32
 		hashFn   HashFunc
-		shards   map[uint32]*idMutexShardImpl
+		shards   []*idMutexShardImpl
```

Comment:
we should keep this as map so we don't need to know the shardCount. It is also a waste to have to allocate all shards for each instance.

Reply:
These are a small number of internal shards, like 4–16, not history shards